File: add_namespace.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <filesystem>

namespace fs = std::filesystem;

// 处理单个.cpp文件，添加namespace XVisual
bool processCppFile(const fs::path& filePath) {
    // 读取文件内容
    std::ifstream inFile(filePath);
    if (!inFile.is_open()) {
        std::cerr << "Cannot open file: " << filePath << std::endl;
        return false;
    }

    std::string content((std::istreambuf_iterator<char>(inFile)),
                        std::istreambuf_iterator<char>());
    inFile.close();

    // 检查是否已经包含namespace XVisual
    if (content.find("namespace XVisual") != std::string::npos) {
        std::cout << "Skipped (already has namespace): " << filePath << std::endl;
        return true;
    }

    // 找到第一个#include之后的位置
    size_t pos = 0;
    size_t lastIncludePos = 0;
    while ((pos = content.find("#include", pos)) != std::string::npos) {
        size_t endLine = content.find('\n', pos);
        lastIncludePos = endLine;
        pos = endLine + 1;
    }

    // 如果找到include，在其后添加namespace开始
    if (lastIncludePos > 0) {
        // 插入namespace XVisual开始
        content.insert(lastIncludePos + 1, "\nnamespace XVisual {\n");
        
        // 在文件末尾添加namespace结束
        if (content.back() == '\n') {
            content += "\n} // namespace XVisual\n";
        } else {
            content += "\n\n} // namespace XVisual\n";
        }
        
        // 写回文件
        std::ofstream outFile(filePath);
        if (!outFile.is_open()) {
            std::cerr << "Cannot write to file: " << filePath << std::endl;
            return false;
        }
        outFile << content;
        outFile.close();
        
        std::cout << "Processed: " << filePath << std::endl;
        return true;
    }

    std::cout << "Skipped (no includes): " << filePath << std::endl;
    return false;
}

int main() {
    std::string basePath = "XVisual/src";
    
    int processedCount = 0;
    int skippedCount = 0;
    
    // 递归遍历src目录
    for (const auto& entry : fs::recursive_directory_iterator(basePath)) {
        if (entry.is_regular_file()) {
            std::string ext = entry.path().extension().string();
            // 只处理.cpp文件，跳过cJSON.c和ui文件
            if (ext == ".cpp") {
                if (processCppFile(entry.path())) {
                    processedCount++;
                } else {
                    skippedCount++;
                }
            }
        }
    }
    
    std::cout << "\n=================================\n";
    std::cout << "Processing completed!\n";
    std::cout << "Processed: " << processedCount << " files\n";
    std::cout << "Skipped: " << skippedCount << " files\n";
    std::cout << "=================================\n";
    
    return 0;
}

File: add_namespace.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
自动为XVisual项目中的所有.cpp文件添加namespace XVisual
"""

import os
import re
from pathlib import Path

def process_cpp_file(file_path):
    """处理单个.cpp文件，添加namespace XVisual"""
    
    # 读取文件内容
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception as e:
        print(f"无法读取文件 {file_path}: {e}")
        return False
    
    # 检查是否已经包含namespace XVisual
    if 'namespace XVisual' in content:
        print(f"跳过（已包含namespace）: {file_path}")
        return True
    
    # 找到最后一个#include的位置
    include_pattern = r'^#include.*?$'
    includes = list(re.finditer(include_pattern, content, re.MULTILINE))
    
    if not includes:
        print(f"跳过（没有include）: {file_path}")
        return False
    
    # 获取最后一个include的结束位置
    last_include = includes[-1]
    last_include_end = last_include.end()
    
    # 在最后一个include后添加namespace开始
    before_namespace = content[:last_include_end]
    after_namespace_start = content[last_include_end:]
    
    # 在文件末尾添加namespace结束
    new_content = before_namespace + "\nnamespace XVisual {\n" + after_namespace_start + "\n} // namespace XVisual\n"
    
    # 写回文件
    try:
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(new_content)
        print(f"已处理: {file_path}")
        return True
    except Exception as e:
        print(f"无法写入文件 {file_path}: {e}")
        return False

def main():
    """主函数：遍历src目录并处理所有.cpp文件"""
    base_path = "XVisual/src"
    
    if not os.path.exists(base_path):
        print(f"错误：目录 {base_path} 不存在")
        return
    
    processed_count = 0
    skipped_count = 0
    
    # 递归遍历src目录
    for root, dirs, files in os.walk(base_path):
        for file in files:
            # 只处理.cpp文件
            if file.endswith('.cpp'):
                file_path = os.path.join(root, file)
                if process_cpp_file(file_path):
                    processed_count += 1
                else:
                    skipped_count += 1
    
    print("\n" + "=" * 50)
    print("处理完成！")
    print(f"已处理: {processed_count} 个文件")
    print(f"跳过: {skipped_count} 个文件")
    print("=" * 50)

if __name__ == "__main__":
    main()